arch/risc-v/src/litex/litex_sdio.c

```diff
@@ -0,0 +1,1450 @@
+
+static sdio_capset_t litex_capabilities(FAR struct sdio_dev_s *dev)
+{
+  sdio_capset_t caps = 0;
+
+  /* LiteSDCard only supports 4-bit bus width */
+
+  caps |= SDIO_CAPS_4BIT_ONLY;
+  caps |= SDIO_CAPS_DMASUPPORTED;
+  caps |= SDIO_CAPS_DMABEFOREWRITE;
+
+  return caps;
+}
+
+/****************************************************************************
+ * Name: litex_status
+ *
+ * Description:
+ *   Get SDIO status.
+ *
+ * Input Parameters:
+ *   dev   - Device-specific state data
+ *
+ * Returned Value:
+ *   Returns a bitset of status values (see litex_dev_s* defines)
+ *
+ ****************************************************************************/
+
+static sdio_statset_t litex_status(FAR struct sdio_dev_s *dev)
+{
+  struct litex_dev_s *priv = (struct litex_dev_s *)dev;
+
+  sdio_statset_t cd = priv->cdstatus;
+  mcinfo("CD Status: %u\n", cd);
+
+  return cd;
+}
+
+/****************************************************************************
+ * Name: litex_widebus
+ *
+ * Description:
+ *   Called after change in Bus width has been selected (via ACMD6).  Most
+ *   controllers will need to perform some special operations to work
+ *   correctly in the new bus mode.
+ *
+ * Input Parameters:
+ *   dev  - An instance of the SDIO device interface
+ *   wide - true: wide bus (4-bit) bus mode enabled
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void litex_widebus(FAR struct sdio_dev_s *dev, bool wide)
+{
+  /* LiteSDCard only supports 4-bit bus width.
+   * Nothing to do here.
+   */
+}
+
+/****************************************************************************
+ * Name: litex_clock
+ *
+ * Description:
+ *   Enable/disable SDIO clocking
+ *
+ * Input Parameters:
+ *   dev  - An instance of the SDIO device interface
+ *   rate - Specifies the clocking to use (see enum sdio_clock_e)
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void litex_clock(FAR struct sdio_dev_s *dev, enum sdio_clock_e rate)
+{
+  uint32_t clk_freq;
+
+  switch (rate)
+    {
+      /* Return early - SDPHY doesnt support clock disabling */
+
+      default:
+      case CLOCK_SDIO_DISABLED:
+        return;
+
+      /* Enable in initial ID mode clocking (<400KHz) */
+
+      case CLOCK_IDMODE:
+        clk_freq = CONFIG_LITEX_IDMODE_FREQ;
+        break;
+
+      /* Enable in MMC normal operation clocking */
+
+      case CLOCK_MMC_TRANSFER:
+        clk_freq = CONFIG_LITEX_MMCXFR_FREQ;
+        break;
+
+      /* SD normal operation clocking (wide 4-bit mode) */
+
+      case CLOCK_SD_TRANSFER_4BIT:
+        clk_freq = CONFIG_LITEX_SD4BIT_FREQ;
+        break;
+    }
+
+  /* Set the new clock frequency along with the clock enable/disable bit */
+
+  uint32_t divider;
+  divider = clk_freq ? litex_get_hfclk() / clk_freq : 256;
```

Comment:
Can we have `#define MAX_DIVIDER 256` and use it in all places?

Reply:
Yep! Good idea. Done